compiler/annotation/expression/block_compiler.rs

```diff
@@ -66,173 +78,195 @@ pub fn compile_expressions<'block, Snapshot: ReadableSnapshot>(
-                }))?;
-            }
-            index.insert(left, expression_binding);
+    for expression_binding in conjunction.constraints().iter().filter_map(|c| c.as_expression_binding()) {
+        let left = expression_binding.left().as_variable().unwrap();
+        if index.insert(left, vec![expression_binding]).is_some() {
+            return Err(Box::new(ExpressionCompileError::MultipleAssignmentsForVariable {
+                variable: context.variable_name(&left),
+                source_span: expression_binding.source_span(),
+            }))?;
         }
     }
+
     for nested in conjunction.nested_patterns() {
         match nested {
             NestedPattern::Disjunction(disjunction) => {
-                for nested_conjunction in disjunction.conjunctions() {
-                    index_expressions(context, nested_conjunction, index)?;
-                }
+                index_expressions_disjunction(context, disjunction, index)?;
             }
             NestedPattern::Negation(negation) => {
-                index_expressions(context, negation.conjunction(), index)?;
+                index_expressions_conjunction(context, negation.conjunction(), index)?;
             }
             NestedPattern::Optional(optional) => {
-                index_expressions(context, optional.conjunction(), index)?;
+                index_expressions_conjunction(context, optional.conjunction(), index)?;
             }
         }
     }
     Ok(())
 }
 
-fn compile_expressions_recursive<'a, Snapshot: ReadableSnapshot>(
-    context: &mut BlockExpressionsCompilationContext<'a, Snapshot>,
-    assigned_variable: Variable,
-    expression_assignments: &HashMap<Variable, &'a ExpressionBinding<Variable>>,
+fn index_expressions_disjunction<'block, Snapshot: ReadableSnapshot>(
+    context: &BlockExpressionsCompilationContext<'_, Snapshot>,
+    disjunction: &'block Disjunction,
+    index: &mut HashMap<Variable, Vec<&'block ExpressionBinding<Variable>>>,
 ) -> Result<(), Box<ExpressionCompileError>> {
-    context.visited_expressions.insert(assigned_variable);
-    let expression = expression_assignments.get(&assigned_variable).unwrap().expression();
-    for variable in expression.variables() {
-        resolve_type_for_variable(context, variable, expression_assignments)?;
-    }
-    let compiled =
-        ExpressionCompilationContext::compile(expression, &context.variable_value_types, context.parameters)?;
-    context.variable_value_types.insert(assigned_variable, compiled.return_type.clone());
-    context.compiled_expressions.insert(assigned_variable, compiled);
+    let mut combined_indices = HashMap::<Variable, Vec<&'block ExpressionBinding<Variable>>>::new();
+    let mut branch_indices = disjunction
+        .conjunctions()
+        .iter()
+        .map(|branch| {
+            let mut branch_index = HashMap::new();
+            index_expressions_conjunction(context, branch, &mut branch_index).map(|_| branch_index)
```

Comment:
Can you add 1 test that makes sure this works across two levels of nesting?
```
match
{ let $x = ... } 
or
{ 
  $y isa person;
  { let $x = ... }
  or 
  { let $x = ... };
};
```
Seems like it should work but is the kind of thing that would easily go wrong